torch/csrc/jit/passes/onnx/fold_if_node.cpp

```diff
@@ -0,0 +1,219 @@
+#include <torch/csrc/jit/passes/onnx/fold_if_node.h>
+#include <torch/csrc/jit/passes/onnx/helper.h>
+#include <torch/csrc/jit/passes/onnx/scalar_type_analysis.h>
+#include <torch/torch.h>
+
+#include <c10/util/Optional.h>
+#include <algorithm>
+
+namespace torch {
+namespace jit {
+
+namespace onnx {
+using namespace ::c10::onnx;
+}
+
+static bool checkIfFold(Node* node, bool dynamic_axes) {
+  if (node->kind() != onnx::If)
+    return false;
+
+  auto cast_node = node->input()->node();
+  // assumes that Cast node is always before If node,
+  // but sometimes that is not the case.
+  if (cast_node->kind() != onnx::Cast)
+    return false;
+  auto prev_node = cast_node->input()->node();
+  
+  Node* compare_node;
+  if (prev_node->kind() == onnx::Not || prev_node->kind() == onnx::Identity) {
+    compare_node = prev_node->input()->node();
+  } else {
+    compare_node = cast_node->input()->node();
+  }
+  if (compare_node->kind() == onnx::If) {
+    return checkIfFold(compare_node, dynamic_axes);
+  }
+  if (compare_node->kind() == onnx::Equal ||
+      compare_node->kind() == onnx::Greater ||
+      compare_node->kind() == onnx::Less) {
+    for (size_t i = 0; i < compare_node->inputs().size(); i++) {
+      if (!(compare_node->inputs()[i]->node()->kind() == onnx::Constant ||
+            compare_node->inputs()[i]->node()->kind() == onnx::Size ||
+            compare_node->inputs()[i]->node()->kind() == onnx::ReduceProd))
+        return false;
+      if (compare_node->inputs()[i]->node()->kind() == onnx::ReduceProd &&
+          dynamic_axes)
+        return false;
+    }
+    return true;
+  } else if (compare_node->kind() == onnx::Constant) {
+    return true;
+  }
+  return false;
+}
+
+static bool constantFoldingValue(Node* node) {
+  auto cast_node = node->input()->node();
+  auto prev_node = cast_node->input()->node();
+  if (prev_node->kind() == onnx::If) {
+    int cond = 1 - (int)constantFoldingValue(prev_node);
+    Block* block = prev_node->blocks()[cond];
+    // we are assuming that the node will be Constant and
+    // that number of block outputs is 1.
+    prev_node = block->outputs()[0]->node();
+  }
+
+  if (prev_node->kind() == onnx::Constant) {
+    const auto val = prev_node->t(attr::value);
+    return at::is_nonzero(val);
+  }
+
+  if (prev_node->kind() == onnx::Identity &&
+      prev_node->input()->node()->kind() == onnx::Constant) {
+    auto val = prev_node->input()->node()->t(attr::value);
+    return at::is_nonzero(val);
+  }
+
+  Node* compare_node;
+  if (prev_node->kind() == onnx::Not) {
+    compare_node = prev_node->input()->node();
+  } else if (cast_node->inputs().size() > 0) {
+    compare_node = cast_node->input()->node();
+  }
```

Comment:
You should probably initialize `compare_node` so that it has a well-defined value on all code paths.

Reply:
yes, you are right. I updated my code. Thanks!